h/services/group_members.py

```diff
@@ -111,17 +115,34 @@ def update_members(self, group, userids):
         for userid in userids_for_removal:
             self.member_leave(group, userid)
 
-    def member_join(self, group, userid) -> GroupMembership:
-        """Add `userid` to the member list of `group`."""
+    def member_join(self, group, userid, roles=None) -> GroupMembership:
+        """
+        Add `userid` to `group` with `roles` and return the resulting membership.
+
+        If `roles=None` it will default to `[GroupMembershipRoles.MEMBER]`.
+
+        If a membership matching `group`, `userid` and `roles` already exists
+        in the DB it will just be returned.
+
+        :raise ConflictError: if a membership already exists with the given
+            group and userid but different roles
+        """
```

Comment:
Previously if you called this method and a matching membership already existed it would just return the existing membership. Now that I've added the `roles` argument it introduces a new case:

It's now possible to call this method when a membership already exists with the matching `group` and `userid` but with different `roles`. In this case we can't create a new membership. Nor do we want to update the existing membership (for example the API view that calls this method doesnt' want it to update existing memberships: a `POST` API should never update an existing resource). So raise an exception in that case.

For backwards-compatibility it will still return a pre-existing membership _if_ it has the same roles.